aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,402 @@
+  }
+
+  finalize change_validator(new_validator: address, rebond_amount: u64) {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 10u8);
+
+    let current_validator: address = validator.get(0u8);
+    assert(new_validator != current_validator);
+
+    let account_balance: u64 = 0u64;
+    let unbonding_balance: u64 = 0u64; // should theoretically be 0
+    let total_balance: u64 = account_balance + unbonding_balance;
+    let pending_withdrawals: u64 = pending_withdrawal.get(0u8);
+    let rebond_balance: u64 = total_balance - pending_withdrawals;
+    assert(rebond_amount <= rebond_balance);
+
+    // Update validator
+    validator.set(0u8, new_validator);
+
+    // Update state
+    state.set(0u8, 0u8);
+  }
+
+  inline calculate_commission_earned(
+    bonded: u64,
+    total_deposits: u64,
+    remaining_at_last_claim: u64,
+    commission_rate: u128
+  ) -> u64 {
+    let total_rewards: u64 = bonded - total_deposits;
+    let rewards_since_last_claim: i64 = total_rewards as i64 - remaining_at_last_claim as i64;
+    let rewards_pos: bool = rewards_since_last_claim > 0i64;
+
+    if (!rewards_pos) {
+      return 0u64;
+    }
+
+    let rewards_unsigned: u128 = rewards_since_last_claim as u128;
+    let new_commission: u128 = rewards_unsigned * commission_rate / PRECISION_UNSIGNED;
+    let new_commission_64: u64 = new_commission as u64;
+    return new_commission_64;
+  }
+
+  transition unbond_commission(amount: u64) {
+    assert_eq(self.caller, ADMIN);
+    credits.aleo/unbond_public(amount);
+
+    return then finalize(amount);
+  }
+
+  finalize unbond_commission(amount: u64) {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 0u8);
+
+    // Calculate commission earned since last claim
+    let bonded: u64 = 0u64;
+    let total_deposits: u64 = net_deposits.get(0u8);
+    let rewards_at_last_claim: u64 = last_commission_claim.get(0u8).remaining_rewards;
+    let commission_rate: u128 = commission_percent.get(0u8);
+    let commission_earned: u64 = calculate_commission_earned(bonded, total_deposits, rewards_at_last_claim, commission_rate);
+
+    assert(commission_earned > 0u64);
+    assert(commission_earned >= amount);
+
+    state.set(0u8, 1u8);
+
+    // Set last_commission_claim
+    let total_rewards: u64 = bonded - total_deposits;
+    let new_claim: commission_claim = commission_claim {
+      height: block.height + UNBONDING_PERIOD,
+      microcredits_claimed: commission_earned,
+      remaining_rewards: total_rewards - amount
+    };
+    last_commission_claim.set(0u8, new_claim);
+  }
+
+  transition claim_commission() {
+    credits.aleo/claim_unbond_public();
+
+    return then finalize();
+  }
+
+  finalize claim_commission() {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 1u8);
+
+    let unbonding_height: u32 = last_commission_claim.get(0u8).height;
+    assert(block.height >= unbonding_height);
```

Comment:
Is this necessary? The claim_commission() transition should fail if the unbonding period has not elapsed.